integration/src/main/java/scripts/testscripts/DatasetPermissionOperations.java

```diff
@@ -0,0 +1,238 @@
+      fail("Expected exception was not thrown");
+    } catch (ApiException e) {
+      assertThat(
+          userDatasetsApi.getApiClient().getStatusCode(),
+          is(HttpStatusCodes.STATUS_CODE_UNAUTHORIZED));
+    }
+  }
+
+  private void testNoPermissions(DatasetsApi userDatasetsApi) throws Exception {
+    UUID snapshotId = createSnapshot(adminUser);
+    var request =
+        new CreateDatasetRequest()
+            .catalogEntry(METADATA)
+            .storageSourceId(snapshotId.toString())
+            .storageSystem(StorageSystem.TDR);
+    // Create a catalog entry as admin and verify the user cannot access data
+    var datasetId = adminCreateDataset(request);
+    try {
+      userDatasetsApi.getDataset(datasetId);
+      fail("Expected exception was not thrown");
+    } catch (ApiException e) {
+      assertThat(
+          userDatasetsApi.getApiClient().getStatusCode(),
+          is(HttpStatusCodes.STATUS_CODE_UNAUTHORIZED));
+    }
+  }
+
+  private void testAdminPermissionsOnUserSnapshot(DatasetsApi userDatasetsApi) throws Exception {
+    UUID snapshotId = createSnapshot(regularUser);
+    var request =
+        new CreateDatasetRequest()
+            .catalogEntry(METADATA)
+            .storageSourceId(snapshotId.toString())
+            .storageSystem(StorageSystem.TDR);
+    // Verify admin can create a dataset on the user snapshot
+    var datasetId = adminCreateDataset(request);
+    // But admin cannot access the underlying user snapshot
+    try {
+      adminSnapshotsApi.retrieveSnapshot(snapshotId, Collections.emptyList());
+      fail("Expected exception was not thrown");
+    } catch (Exception e) {
+      assertThat(
+          adminSnapshotsApi.getApiClient().getStatusCode(),
+          is(HttpStatusCodes.STATUS_CODE_UNAUTHORIZED));
+    }
+  }
+
+  private UUID adminCreateDataset(CreateDatasetRequest request) throws Exception {
+    var datasetId = adminDatasetsApi.createDataset(request).getId();
+    assertThat(adminDatasetsApi.getApiClient().getStatusCode(), is(HttpStatusCodes.STATUS_CODE_OK));
+    datasetIds.add(datasetId);
+    return datasetId;
+  }
+
+  private UUID createSnapshot(TestUserSpecification user) throws Exception {
+    var datarepoClient = new DatarepoClient(server, user);
+    var snapshotsApi = new SnapshotsApi(datarepoClient);
+    var request =
+        new SnapshotRequestModel()
+            .name("catalog_integration_test_" + System.currentTimeMillis())
+            .description("catalog test snapshot")
+            .profileId(defaultProfileId)
+            .addContentsItem(
+                new SnapshotRequestContentsModel()
+                    .datasetName(TEST_DATASET_NAME)
+                    .mode(SnapshotRequestContentsModel.ModeEnum.BYFULLVIEW));
+    var createJobId = snapshotsApi.createSnapshot(request).getId();
+
+    var jobApi = new JobsApi(datarepoClient);
+    JobModel job;
+    do {
+      job = jobApi.retrieveJob(createJobId);
```

Comment:
Should we sleep here to not burn CPU and network requests?

Reply:
That's a good point, this is code copied from `SnapshotDatasetOperations.setup()` but the same is true there; the code should be sleeping as we know it takes ~60s for this to complete. Maybe a 10 second sleep would be good?